boltons/ioutils.py

```diff
@@ -0,0 +1,322 @@
+
+    Example::
+
+        >>> from boltons import ioutils
+        >>> with ioutils.SpooledBytesIO() as f:
+        ...     f.write(b"Happy IO")
+        ...     _ = f.seek(0)
+        ...     isinstance(f.getvalue(), ioutils.binary_type)
+        True
+    """
+
+    def read(self, n=-1):
+        return self.buffer.read(n)
+
+    def write(self, s):
+        if not isinstance(s, binary_type):
+            raise TypeError("{0} expected, got {1}".format(
+                binary_type.__name__,
+                type(s).__name__
+            ))
+
+        self.buffer.write(s)
+        if self.tell() >= self._max_size:
+            self.rollover()
+
+    def readline(self, length=None):
+        return self.buffer.readline(length)
+
+    def readlines(self, sizehint=0):
+        return self.buffer.readlines(sizehint)
+
+    def rollover(self):
+        """Roll the StringIO over to a TempFile"""
+        if not self._rolled:
+            tmp = TemporaryFile()
+            pos = self._file.tell()
+            self.__buffer.seek(0)
+            tmp.write(self.__buffer.read())
+            tmp.seek(pos)
+            self.__buffer.close()
+            self.__buffer = tmp
+
+    @property
+    def _rolled(self):
+        return not isinstance(self.buffer, BytesIO)
+
+    @property
+    def buffer(self):
+        try:
+            return self.__buffer
+        except AttributeError:
+            self.__buffer = BytesIO()
+        return self.__buffer
+
+
+class SpooledStringIO(SpooledIOBase):
+    """
+    SpooledStringIO is a spooled file-like-object that only accepts unicode
+    values. On Python 2.x this means the 'unicode' type and on Python 3.x this
+    means the 'str' type. Values are accepted as unicode and then coerced into
+    utf-8 encoded bytes for storage. On retrieval, the values are returned as
+    unicode.
+
+    Example::
+
+        >>> from boltons import ioutils
+        >>> with ioutils.SpooledStringIO() as f:
+        ...     f.write(u"\u2014 Hey, an emdash!")
+        ...     _ = f.seek(0)
+        ...     isinstance(f.read(), ioutils.text_type)
+        True
+
+     """
+
+    def read(self, n=-1):
+        return self.buffer.read(n).decode('utf-8')
+
+    def write(self, s):
+        if not isinstance(s, text_type):
+            raise TypeError("{0} expected, got {1}".format(
+                text_type.__name__,
+                type(s).__name__
+            ))
+
+        self.buffer.write(s.encode('utf-8'))
+        if self.tell() >= self._max_size:
+            self.rollover()
+
+    def readline(self, length=None):
+        return self.buffer.readline(length).decode('utf-8')
+
+    def readlines(self, sizehint=0):
+        return [x.decode('utf-8') for x in self.buffer.readlines(sizehint)]
+
+    @property
+    def buffer(self):
+        try:
+            return self.__buffer
+        except AttributeError:
+            self.__buffer = EncodedFile(BytesIO(), data_encoding='utf-8')
```

Comment:
@induane There'd be no need for `EncodedFile` if `TemporaryFile` could be made to handle encoding.  But it that relies on behavior in Python 3 so it's irrelevant.

It turns out, though, that `EncodedFile` is useless here.  That's because it translates a byte string in one encoding to a bye string in another encoding, and stores the latter in a file.  So this is an error:

```python
codecs.EncodedFile(io.BytesIO(), data_encoding='utf-8').write(u'\u2022')
```

Because encoded file wants bytes as input.

I think what you actually want is `codecs.StreamReaderWriter`.  That's what `codecs.open` uses.  Here's how you might replace `EncodedFile` with it:

```python
self._mem_buf = BytesIO()
codec_info = codecs.lookup('utf-8')
self._buffer = codecs.StreamReaderWriter(self._mem_buf,
                                         codecs_info.reader,
                                         codecs_info.writer)
```

With that change you won't need to utf-8 encode any of the reads or the writes to `self.buffer`.  However, you'll need to query the in-memory buffer to find out the length of the encoded data before a rollover.  e.g.:

```python
@property
def len(self):
    if self._rolled:
        ...
    else:
        return len(self._mem_buf.getvalue())
```

You'll also need to implement `tell()` in the same way, as `StreamReaderWriter` has a `seek` but no `tell` method.

If all that is too much work, you can just get rid of `EncodedFile` entirely.  The advantage there would be that the `rollover` implementation between the two instances would become identical.

One last comment:  it appears that you always want the number of bytes of *encoded* data, not the number of code points, so that writing `u'\u2022'` will increase the length by 3, not 1.  If it were possible to return the number of code points instead (like `io.StringIO` would), then you could avoid encoding when writing to an in memory buffer by using `io.StringIO`.  That doesn't appear to be the desired API, though.